File: java/modules/basic/src/main/java/io/v6d/modules/basic/arrow/RecordBatch.java
/** Copyright 2020-2023 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.v6d.modules.basic.arrow;

import com.google.common.base.Objects;
import io.v6d.core.client.ds.Object;
import io.v6d.core.client.ds.ObjectFactory;
import io.v6d.core.client.ds.ObjectMeta;
import io.v6d.modules.basic.arrow.util.ObjectResolver;
import io.v6d.modules.basic.columnar.ColumnarData;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import lombok.val;
import org.apache.arrow.vector.FieldVector;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class RecordBatch extends Object {
    private static final Logger logger = LoggerFactory.getLogger(RecordBatch.class);

    private final VectorSchemaRoot batch;
    private ObjectResolver resolver;

    public static void instantiate() {
        Schema.instantiate();
        BooleanArray.instantiate();
        Int8Array.instantiate();
        Int16Array.instantiate();
        Int32Array.instantiate();
        Int64Array.instantiate();
        FloatArray.instantiate();
        DoubleArray.instantiate();
        StringArray.instantiate();
        VarBinaryArray.instantiate();
        LargeStringArray.instantiate();
        NullArray.instantiate();
        DateArray.instantiate();
        TimestampArray.instantiate();
        DecimalArray.instantiate();
        ListArray.instantiate();
        StructArray.instantiate();
        MapArray.instantiate();
        ObjectFactory.getFactory().register("vineyard::RecordBatch", new RecordBatchResolver());
    }

    public RecordBatch(final ObjectMeta meta, Schema schema, List<FieldVector> vectors, int nrow) {
        super(meta);
        this.batch = new VectorSchemaRoot(schema.getSchema(), vectors, nrow);
        resolver = new ObjectResolver();
    }

    public VectorSchemaRoot getBatch() {
        return batch;
    }

    public long getRowCount() {
        return batch.getRowCount();
    }

    public long getColumnCount() {
        return batch.getFieldVectors().size();
    }

    public void setResolver(ObjectResolver resolver) {
        this.resolver = resolver;
    }

    public ColumnarData[] columar() {
        List<FieldVector> vectors = batch.getFieldVectors();
        ColumnarData[] columnarData = new ColumnarData[vectors.size()];
        for (int i = 0; i < vectors.size(); i++) {
            columnarData[i] = new ColumnarData(vectors.get(i), resolver);
        }
        return columnarData;
    }

    @Override
    public boolean equals(java.lang.Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        RecordBatch that = (RecordBatch) o;
        return Objects.equal(batch, that.batch);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(batch);
    }
}

class RecordBatchResolver extends ObjectFactory.Resolver {
    @Override
    public Object resolve(final ObjectMeta meta) {
        val schema = (Schema) new SchemaResolver().resolve(meta.getMemberMeta("schema_"));
        val ncol = meta.getIntValue("column_num_");
        val nrow = meta.getIntValue("row_num_");

        val vectors =
                IntStream.range(0, meta.getIntValue("__columns_-size"))
                        .mapToObj(
                                index -> {
                                    val column = meta.getMemberMeta("__columns_-" + index);
                                    return ((Array) ObjectFactory.getFactory().resolve(column))
                                            .getArray();
                                })
                        .collect(Collectors.toList());
        return new RecordBatch(meta, schema, vectors, nrow);
    }
}

File: python/vineyard/io/recordbatch.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright 2020-2023 Alibaba Group Holding Limited.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

''' This module exposes support for RecordBatchStream.
'''

import contextlib
import json
from typing import Any
from typing import Dict
from typing import Optional

from vineyard._C import ObjectMeta
from vineyard.core import context
from vineyard.io.stream import BaseStream


class RecordBatchStream(BaseStream):
    def __init__(self, meta: ObjectMeta, params: Optional[Dict[str, Any]] = None):
        super().__init__(meta)
        self._params = params

    @property
    def params(self):
        return self._params

    @staticmethod
    def new(
        client,
        params: Optional[Dict[str, Any]] = None,
        meta: Optional[ObjectMeta] = None,
    ) -> "RecordBatchStream":
        if meta is None:
            meta = ObjectMeta()
        meta['typename'] = 'vineyard::RecordBatchStream'
        if params is None:
            params = dict()
        meta['params_'] = params
        meta = client.create_metadata(meta)
        client.create_stream(meta.id)
        return RecordBatchStream(meta, params)


def recordbatch_stream_resolver(obj, resolver):  # pylint: disable=unused-argument
    meta = obj.meta
    if 'params_' in meta:
        params = json.loads(meta['params_'])
    else:
        params = dict
    return RecordBatchStream(meta, params)


def register_recordbatch_stream_types(_builder_ctx, resolver_ctx):
    if resolver_ctx is not None:
        resolver_ctx.register(
            'vineyard::RecordBatchStream', recordbatch_stream_resolver
        )


@contextlib.contextmanager
def recordbatch_stream_context():
    with context() as (builder_ctx, resolver_ctx):
        register_recordbatch_stream_types(builder_ctx, resolver_ctx)
        yield builder_ctx, resolver_ctx
